### PR TITLE
feat(multi-target-repo): add support for multi-target repositories

### PR DESCRIPTION
**Pull request by qtpeters**
## Summary

Design and implement a way to extract every target (build file) from a given repository using the GitHub Client

### Added

- Git Service to handle Git Client specific operations.
- Syft Service to manage Syft actions.

### Changed

- Refactored GitHub Provider Sync service to handle searching for build files associated to a particular cataloger
- Now GitHub Provider will create an SBOM for every build target specified in the configuration

## How to test

- Start test environment
- Run `/target/debug/harbor ingest --provider github harbor-test-org --debug`
- Verify that:
  - 18 `primary` SBOMs are produced and are in MongoDB 
  - Provider produces 1 SBOM for: `https://github.com/harbor-test-org/java-repo`. Look in MongoDB for purl like:
    - pkg:**java-pom**:harbor-test-org/**java-repo**@|commit-hash|
    - In the example above, the important pieces of data are: 
      - `java-pom`: The cataloger used to create the SBOM.
      - `java-repo`: The name of the repository the SBOM was created from
      - `harbor-test-org`: is also important because it is the organization the repos are in  
  - Provider produces 11 SBOMS from the java multi repo like:
    - pkg:**java-pom**:harbor-test-org/**java-multi-module**@|commit-hash|#**lib/api-docs**
    - In the example above, also look to see that the file location paths are unique
  - Provider produces 2 SBOMs from the `ror-repo-with-gemfile` repo.  One from Ruby and one from JavaScript:
    - pkg:**ruby-gemfile**:harbor-test-org/**ror-repo-with-gemfile**@|commit-hash|
    - pkg:**javascript-package**:harbor-test-org/**ror-repo-with-gemfile**@|commit-hash|
  - Provider produces 1 SBOM from the `python-project-with-requirements-txt` repo:
    - pkg:**python-index**:harbor-test-org/**python-project-with-requirements-txt**@|commit-hash|
  - Provider produces 1 SBOM from the `javascript-repo-correct` repo:
    - pkg:**javascript-package**:harbor-test-org/**javascript-repo-correct**@|commit-hash|
  - Provider produces only 1 SBOM from `javascript-repo-node-modules-checked-in`.  This is important, because the 
    node-modules folder is checked in and populated with dependencies.  This tests the ability for the Provider to avoid
    processing build targets when they are under specified directories.  This feature is mandatory because I have 
    identified at least one case of this in the CMS repos. 
    - pkg:**javascript-package**:harbor-test-org/**javascript-repo-node-modules-checked-in**@|commit-hash|
  - Provider produces 1, `no-cataloger` SBOM in the `terraform-only` repo as there are no build targets.
    - pkg:**no-cataloger**:harbor-test-org/**terraform-only**@|commit-hash|

### Points of interest:
- Forked repos are not processed by the GitHub Provider. They must be repos that are created in a given organization to be processed. If we want to process Forks as well, we will have to refactor.
- This iteration of the GitHub provider supports: Java, Javascript, Python, Ruby and Hail Mary repos with no build target
- Currently, the configuration for the catalogers and build files resides in `sdk/core/src/services/github/service.rs`.  This config should be encoded in JSON and moved somewhere that makes sense

### CODE REVIEW

File: sdk/core/src/tasks/sboms/github/sync.rs [hunk 1]
```diff
@@ -34,20 +63,22 @@ pub struct SyncTask {
 impl TaskProvider for SyncTask {
     /// Builds the Packages Dependencies, Purls, and Unsupported from the GitHub API.
     async fn run(&self, task: &mut Task) -> Result<HashMap<String, String>, Error> {
-        let mut errors = HashMap::new();
-
         let mut repos: Vec<Repo> = match self.github.get_repos().await {
             Ok(value) => value,
             Err(err) => {
-                println!("Error attempting to get the repos: {}", err);
+                println!("Error attempting to get the repos: {:#?}", err);
```
Comment (DerekStrickland) [inline sdk/core/src/tasks/sboms/github/sync.rs:69]: Do we want to leave this debug in here on purpose?  My perennial fear around this topic is that we might end up leaking things into logs that we can't predict.

Comment (DerekStrickland) [inline sdk/core/src/tasks/sboms/github/sync.rs:69]: Do we want to leave this debug in here on purpose?  My perennial fear around this topic is that we might end up leaking things into logs that we can't predict.

File: sdk/core/src/tasks/sboms/github/sync.rs [hunk 2]
```diff
@@ -121,53 +124,77 @@ impl Service<Task> for SyncTask {
...
+        let mut errors: HashMap<String, String> = HashMap::new();
+
+        let xref = Xref {
+            kind: XrefKind::Product,
+            map: HashMap::new(),
+        };
+
+        match self
+            .sboms
+            .ingest(sbom.as_str(), None, SbomProviderKind::GitHub, xref, task)
```
Comment (DerekStrickland) [inline sdk/core/src/tasks/sboms/github/sync.rs:142]: I'm adding the HarborSyft kind to main and it should be there by tomorrow AM.

File: sdk/core/src/tasks/sboms/github/sync.rs [hunk 3]
```diff
@@ -223,71 +252,120 @@ impl SyncTask {
...
 
+        let last_hash: &String = &repo.last_hash;
+
         println!(
             "==> Comparing Repo({}) to MongoDB({:#?})",
             last_hash, document.last_hash
         );
 
-        return if *last_hash != document.last_hash.unwrap() {
+        if *last_hash != document.last_hash.unwrap() {
```
Comment (DerekStrickland) [inline sdk/core/src/tasks/sboms/github/sync.rs:277]: Does this get refactored with the move to default branch name? Do you think we should merge this, and do that change in another PR, or just update this PR?  To me it depends on how long that would take.

File: sdk/core/src/tasks/sboms/github/sync.rs [hunk 4]
```diff
@@ -223,71 +252,120 @@ impl SyncTask {
...
+                .clone_repo(url.as_str())
+                .map_err(|err| Error::GitHub(format!("Error attempting clone repo: {:#?}", err)))?;
+
+            let syft = Syft::new(clone_path.clone());
+
+            let mut syft_results: Vec<String> = vec![];
+
+            let mut total_build_targets = 0;
+
+            for (cataloger, build_targets) in cataloger_to_build_file_map {
```
Comment (DerekStrickland) [inline sdk/core/src/tasks/sboms/github/sync.rs:291]: Looking at the error handling here, am I correct that if one of the build targets fails the whole repo will fail?  Is that what we want?  I truly don't know, but I'm wondering if it should try to be more fault tolerant like the out task loop.

File: sdk/platform/src/filesystem/mod.rs [hunk 5]
```diff
@@ -1,5 +1,22 @@
 use crate::Error;
 use regex::Regex;
+use crate::str::get_random_string;
+
+/// Generates a valid location for operations on data
+pub fn get_tmp_location() -> String {
+    format!("/tmp/harbor-debug/{}", get_random_string())
```
Comment (DerekStrickland) [inline sdk/platform/src/filesystem/mod.rs:7]:
```suggestion
    format!("/tmp/harbor/{}", get_random_string())
```

**Review by DerekStrickland — COMMENTED** [review]
Sweet job.  Let's connect tomorrow morning on getting it merged. I think I still have some questions and want to make sure we write down follow up work that we'll do later vs. things we do now.

**Comment by qtpeters** [comment]
Let's do it!

File: sdk/core/src/services/github/service.rs [hunk 6]
```diff
@@ -1,28 +1,27 @@
...
-use std::path::Path;
+use platform::git::Service as Git;
+use platform::str::get_random_string;
 
 /// Definition of the GitHubProvider
 #[derive(Debug)]
 pub struct GitHubService {
     org: String,
-    client: GitHubClient,
+    github_client: GitHubClient,
```
[the remark this answers is not in the document]
Comment (qtpeters) [reply sdk/core/src/services/github/service.rs:10]: Good news, I changed that variable name when I had two clients that I had to disambiguate.  That variable can go back to being just `client` now.

File: sdk/core/src/services/github/service.rs [hunk 7]
```diff
@@ -67,31 +68,42 @@ impl GitHubService {
...
+    fn clone_path(&self, url: &str, rand: &String) -> Result<String, Error> {
         let repo_name = url
             .split('/')
             .collect::<Vec<&str>>()
             .pop()
             .unwrap()
             .replace(".git", "");
 
-        Ok(format!("/tmp/harbor-debug/{}/{}", hash, repo_name))
+        Ok(format!("/tmp/harbor-debug/{}/{}", rand, repo_name))
```
[the remark this answers is not in the document]
Comment (qtpeters) [reply sdk/core/src/services/github/service.rs:100]: I think we need to have this value somewhere that every test can get to it.  Look at all of these:

image

File: sdk/core/src/services/syft/mod.rs [hunk 8]
```diff
@@ -0,0 +1,368 @@
...
+                let mut owned_purl = purl;
+                owned_purl.push_str(format!("#{}", value).as_str());
+                owned_purl
+            }
+        }
+    }
+}
+
+/// Syft Service
+pub(crate) struct Service {
```
[the remark this answers is not in the document]
Comment (qtpeters) [reply sdk/core/src/services/syft/mod.rs:63]: I was thinking that services could be defined the same way.  Then, in the future, should we need to add a trait to all Harbor services, then we would have them all defined as structs rather than a simple grouping of functions.  Also, it's nice for new developers to think "If I need to use a service from core, then I know what to look for and have a basic understanding of how they are used due to convention".  That's my take.

File: sdk/core/src/tasks/sboms/github/sync.rs (hunk 1 above)
Comment (qtpeters) [reply sdk/core/src/tasks/sboms/github/sync.rs:69]: No, I can change that back.  Good thinking.

File: sdk/core/src/tasks/sboms/github/sync.rs (hunk 2 above)
Comment (qtpeters) [reply sdk/core/src/tasks/sboms/github/sync.rs:142]: Changed code to reflect this.

File: sdk/core/src/tasks/sboms/github/sync.rs (hunk 3 above)
Comment (qtpeters) [reply sdk/core/src/tasks/sboms/github/sync.rs:277]: I'm unclear about this, however I have distilled it in trello

File: cli/src/commands/ingest/github.rs [hunk 9]
```diff
@@ -77,3 +77,30 @@ pub(crate) async fn execute(args: &IngestArgs) -> Result<(), Error> {
...
 }
+
+#[cfg(test)]
+mod test {
+    use crate::commands::ingest::github::{execute, GitHubArgs};
+    use crate::commands::ingest::{IngestArgs, IngestionProviderKind};
+    use crate::Error;
+
+    #[tokio::test]
+    #[ignore = "manual_debug_test"]
```
[the remark this answers is not in the document]
Comment (qtpeters) [reply cli/src/commands/ingest/github.rs]: Removed manual run requirement

File: sdk/core/src/tasks/sboms/github/sync.rs (hunk 4 above)
Comment (qtpeters) [reply sdk/core/src/tasks/sboms/github/sync.rs:298]: Made error handling fault tolerant

File: sdk/platform/src/filesystem/mod.rs (hunk 5 above)
Comment (qtpeters) [reply sdk/platform/src/filesystem/mod.rs:7]: I have a note to address this

File: sdk/core/src/services/syft/mod.rs [hunk 10]
```diff
@@ -0,0 +1,368 @@
...
+   => name:       the name of the package. Required.
+
+   => version:    the version of the package. Optional.
+
+   => qualifiers: extra qualifying data for a package such as an OS, architecture, a distro, etc.
+                  Optional and type-specific.
+
+   => subpath:    extra subpath within a package, relative to the package root. Optional.
+*/
+fn generate_stopgap_purl(
```
[the remark this answers is not in the document]
Comment (qtpeters) [reply sdk/core/src/services/syft/mod.rs:37]: TODO added

File: sdk/core/src/services/syft/mod.rs [hunk 11]
```diff
@@ -0,0 +1,368 @@
...
+        };
+
+        let output: String = String::from_utf8_lossy(&output.stdout).to_string();
+        let mut json: Value =
+            serde_json::from_str(&output).map_err(|e| Error::SyftError(e.to_string()))?;
+
+        // Access the nested "component" object
+        if let Some(metadata) = json.get_mut("metadata").and_then(Value::as_object_mut) {
+            if let Some(component) = metadata.get_mut("component").and_then(Value::as_object_mut) {
+                component.insert(
```
[the remark this answers is not in the document]
Comment (qtpeters) [reply sdk/core/src/services/syft/mod.rs:163]: In Syft?  Always none remember? We can't get Syft to generate a purl for any of the CMS repos because is can't tell what those values might be.  I experimented a lot with Syft trying to get it to produce one and I can't even find anything in the documentation that says Syft can do such a thing.

File: sdk/core/src/services/syft/mod.rs [hunk 12]
```diff
@@ -0,0 +1,368 @@
...
+
+        if output.stdout.is_empty() {
+            return Err(Error::SyftError("syft generated empty SBOM".to_string()));
+        };
+
+        let output: String = String::from_utf8_lossy(&output.stdout).to_string();
+        let mut json: Value =
+            serde_json::from_str(&output).map_err(|e| Error::SyftError(e.to_string()))?;
+
+        // Access the nested "component" object
```
[the remark this answers is not in the document]
Comment (qtpeters) [reply sdk/core/src/services/syft/mod.rs:160]: Noted and recorded for planning

File: sdk/core/src/services/syft/mod.rs [hunk 13]
```diff
@@ -0,0 +1,368 @@
...
+            })?
+        }
+
+        Ok(serde_json::to_string(&json).unwrap())
+    }
+}
+
+/// Errors specific to Syft operations.
+#[derive(Error, Debug)]
+pub enum Error {
```
[the remark this answers is not in the document]
Comment (qtpeters) [reply sdk/core/src/services/syft/mod.rs:190]: I think I do.  Using the macros, it's easy to properly map them from smaller, more specific errors.  I just think we were learning thiserror while implementing and now that we understand the API better, I think we need to review all the errors and come up with a good way to do this properly.

File: sdk/core/src/services/syft/mod.rs [hunk 14]
```diff
@@ -0,0 +1,368 @@
...
+        if output.stdout.is_empty() {
+            return Err(Error::SyftError("syft generated empty SBOM".to_string()));
+        };
+
+        let output: String = String::from_utf8_lossy(&output.stdout).to_string();
+        let mut json: Value =
+            serde_json::from_str(&output).map_err(|e| Error::SyftError(e.to_string()))?;
+
+        // Access the nested "component" object
+        if let Some(metadata) = json.get_mut("metadata").and_then(Value::as_object_mut) {
```
[the remark this answers is not in the document]
Comment (qtpeters) [reply sdk/core/src/services/syft/mod.rs:161]: I've recorded this for planning

File: sdk/core/src/tasks/sboms/github/mod.rs [hunk 15]
```diff
@@ -1,7 +1,36 @@
...
 pub use sync::*;
 
+/// Temporary way to associate build files with Syft catalogers.  Syft
+/// can use the same cataloger for multiple types of files, that is the reason
+/// we need to use a vector to manage them.
+pub fn get_cataloger_to_build_target_map() -> HashMap<String, Vec<String>> {
+    let mut map = HashMap::new();
+
+    // Add java pom.xml
+    map.insert("java-pom".to_string(), vec!["pom.xml".to_string()]);
```
[the remark this answers is not in the document]
Comment (qtpeters) [reply sdk/core/src/tasks/sboms/github/mod.rs:14]: Yes!  We will be joining forces as we discussed this morning!

File: sdk/core/src/tasks/sboms/github/sync.rs [hunk 16]
```diff
@@ -6,20 +6,49 @@ use crate::entities::xrefs::{Xref, XrefKind};
...
-// scheme:type/namespace/name@version?qualifiers#subpath
-fn generate_stopgap_purl(full_name: String, commit_hash: String) -> String {
-    format!("pkg:harbor:{}@{}", full_name, commit_hash)
+/// This function make sure that if there are certain directories in
+/// the build target path, we should skip processing it as it is a dependency
+/// rather than a primary target.  The immediate first example is node_modules.
+/// node_modules is not supposed to be checked in, however at least one
+/// example has been found and none of the packages under that directory
+/// should be processed as primary build targets.
+fn location_under_ignored_dir(build_target_location: String) -> bool {
```
[the remark this answers is not in the document]
Comment (qtpeters) [reply sdk/core/src/tasks/sboms/github/sync.rs:42]: And target for java projects as well. Instead of a TODO, I added this to my personal backlog and it will become a jira ticket

File: sdk/core/src/tasks/sboms/github/sync.rs (hunk 1 above)
Comment (qtpeters) [reply sdk/core/src/tasks/sboms/github/sync.rs]: Removed debug :+1:

**Review by DerekStrickland — APPROVED** [review]
:shipit:
